vertical-pod-autoscaler/pkg/recommender/routines/safety_margin_modifier_post_processor.go

```diff
@@ -0,0 +1,221 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package routines
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+	"k8s.io/klog/v2"
+)
+
+type SafetyMarginModifierPostProcessor struct {
+	DefaultSafetyMarginFactor float64
+}
+
+type modifierFunction int64
+
+const (
+	undefined modifierFunction = iota
+	linear
+	log
+	affine
+	exponential
+)
+
+type safetyMarginFunction struct {
+	Function   modifierFunction
+	Parameters []float64
+}
+
+type safetyMarginModifier map[v1.ResourceName]safetyMarginFunction
+
+const (
+	vpaPostProcessorSafetyMarginModifierSuffix = "_safetyMarginModifier"
+)
+
+var _ RecommendationPostProcessor = &SafetyMarginModifierPostProcessor{}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *SafetyMarginModifierPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, policy *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	updatedRecommendation := recommendation.DeepCopy()
+	modifiers := readSafetyMarginModifierFromVPAAnnotations(vpa)
+
+	for _, containerRec := range updatedRecommendation.ContainerRecommendations {
+		modifier, found := modifiers[containerRec.ContainerName]
+
+		if !found {
+			continue
+		}
+		klog.Infof("Apply safety marging modifier on container %s in vpa %s/%s", containerRec.ContainerName, vpa.ID.Namespace, vpa.ID.VpaName)
+
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.LowerBound, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.Target, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.UpperBound, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.UncappedTarget, modifier)
+	}
+
+	return updatedRecommendation
+}
+
+func (r *SafetyMarginModifierPostProcessor) applyModifier(vpa *model.Vpa, name string, containerRec v1.ResourceList, modifier safetyMarginModifier) {
+	for resourceName, rec := range containerRec {
+		resourceModifier, found := modifier[resourceName]
+		if !found {
+			continue
+		}
+
+		var newRec float64
+
+		// CPU requests are expressed in milli cores, which impacts the modification brough by
+		// the Log and Exponential modifiers. If the resource is CPU, scale it by 1000.
+		recCopy := rec.DeepCopy()
+		if resourceName == v1.ResourceCPU {
+			recCopy.SetScaled(rec.MilliValue(), 0)
+		}
+
+		switch resourceModifier.Function {
+		case linear:
+			if len(resourceModifier.Parameters) != 1 {
+				klog.Errorf("Skipping %s safety margin modifier in vpa %s/%s: linear modifier requires 1 parameter, %d given", resourceName, name, vpa.ID.Namespace, vpa.ID.VpaName, len(resourceModifier.Parameters))
```

Comment:
maybe return an error instead and print it in the caller ? It would be great to have processors errors eventually propagating to events

Reply:
Sounds good. I'll returns errors up to the `Process` function and klog.Errorf there. Once events can be propagated outisde of post processors, we'll just need to return them instead logging